internal/middlewares/access_managment.go

```diff
@@ -0,0 +1,180 @@
+		acceptableRoles: roles,
+		fullAccessRoles: fullAccessRoles,
+		accessMap:       accessMap,
+	}
+}
+
+type Path struct {
+	PathAsString string
+	PathAsRegexp *regexp.Regexp
+}
+
+func (p *Path) IsMatch(st string) bool {
+	if p.PathAsString != "" {
+		return p.PathAsString == st
+	}
+	return p.PathAsRegexp.MatchString(st)
+}
+
+func NewPath(path string) Path {
+	if strings.Contains(path, "{") {
+		first := strings.Index(path, "{")
+		second := strings.Index(path, "}")
+		path = path[:first] + "(.*)" + path[second+1:]
+		return Path{
+			PathAsRegexp: regexp.MustCompile(path),
+		}
+	} else {
+		return Path{
+			PathAsString: path,
+		}
+	}
+}
+
+func endpointConversion(path string) string {
+	return fmt.Sprintf("/api%s", path) //TODO: remove hardcode
+}
+
+// AddNewAccess adds new access to the access manager. Returns true if access was added, false if access was not added
+func (a *AccessManager) AddNewAccess(role, method, path string) (bool, error) {
+	if !utils.IsValueInList(role, a.acceptableRoles) {
+		return false, errors.New(fmt.Sprintf("role %s is not in the list of acceptable roles", role))
+	}
+	if utils.IsValueInList(role, a.fullAccessRoles) {
+		return false, nil
+	}
+	paths, ok := a.endpoints[method]
+	if ok {
+		if !utils.IsValueInList(path, paths) {
+			return false, errors.New(fmt.Sprintf("path %s is not in the list of existing endpoints", path))
+		}
+
+		endpointsByRole, ok := a.accessMap[role]
+		if !ok {
+			endpointsByRole = make(map[string][]Path)
+			a.accessMap[role] = endpointsByRole
+		}
+		pathToUpdate, ok := endpointsByRole[method]
+		if !ok {
+			pathToUpdate = []Path{
+				NewPath(endpointConversion(path)),
+			}
+			endpointsByRole[method] = pathToUpdate
+			a.accessMap[role] = endpointsByRole
+			return true, nil
+		}
+		if !utils.IsValueInList(NewPath(path), pathToUpdate) {
+			pathToUpdate = append(pathToUpdate, NewPath(endpointConversion(path)))
+			endpointsByRole[method] = pathToUpdate
+			a.accessMap[role] = endpointsByRole
+			return true, nil
+		}
+		return false, nil
+
+	}
+	return false, errors.New(fmt.Sprintf("method %s is not in the list of existing endpoints", method))
+}
+
+// HasAccess checks if role has access to the endpoint
+func (a *AccessManager) HasAccess(role, method, path string) bool {
+	if utils.IsValueInList(role, a.fullAccessRoles) {
+		return true
+	}
+	paths, ok := a.accessMap[role][method]
+	if ok {
+		for _, p := range paths {
+			if p.IsMatch(path) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (a *AccessManager) Middleware() func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			tokenString := r.Header.Get("Authorization")
+			log, err := logger.Get()
+			if err != nil {
+				w.Write([]byte("error getting logger"))
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+			log.Info(fmt.Sprintf("token %s", tokenString))
+			if tokenString == "" {
```

Comment:
If this middleware applied to all endpoints?
We have couple of unauthorized ones.

Reply:
Yeah. This middleware is applied to all endpoints. But it will ignore endpoints without authorization.